File: app/src/main/java/org/alkaline/taskbrain/dsl/runtime/BuiltinRegistry.kt
package org.alkaline.taskbrain.dsl.runtime

import org.alkaline.taskbrain.dsl.builtins.ArithmeticFunctions
import org.alkaline.taskbrain.dsl.builtins.CharacterConstants
import org.alkaline.taskbrain.dsl.builtins.DateFunctions
import org.alkaline.taskbrain.dsl.builtins.NoteFunctions
import org.alkaline.taskbrain.dsl.builtins.PatternFunctions

/**
 * A builtin function that can be called from the DSL.
 *
 * @property name The function name used to call it
 * @property isDynamic True if this function can return different results on each call
 *                     (e.g., `now`, `date`, `time`). Static functions always return the
 *                     same result for the same inputs (e.g., `iso8601`, `qt`).
 * @property call The implementation that takes arguments and an environment
 */
data class BuiltinFunction(
    val name: String,
    val isDynamic: Boolean = false,
    val call: (args: Arguments, env: Environment) -> DslValue
)

/**
 * Registry of all builtin functions available in the DSL.
 *
 * Milestone 3: Adds arithmetic functions.
 * Milestone 5: Adds note functions (find).
 */
object BuiltinRegistry {
    private val functions = mutableMapOf<String, BuiltinFunction>()

    init {
        // Register all builtin modules
        DateFunctions.register(this)
        CharacterConstants.register(this)
        ArithmeticFunctions.register(this)
        PatternFunctions.register(this)
        NoteFunctions.register(this)
    }

    /**
     * Register a builtin function.
     */
    fun register(function: BuiltinFunction) {
        functions[function.name] = function
    }

    /**
     * Look up a function by name.
     * @return The function, or null if not found
     */
    fun get(name: String): BuiltinFunction? = functions[name]

    /**
     * Check if a function exists.
     */
    fun has(name: String): Boolean = functions.containsKey(name)

    /**
     * Check if a function is dynamic (can return different results each call).
     */
    fun isDynamic(name: String): Boolean = functions[name]?.isDynamic ?: false

    /**
     * Get all registered function names.
     */
    fun allNames(): Set<String> = functions.keys.toSet()
}

File: web/src/dsl/runtime/BuiltinRegistry.ts
import type { DslValue } from './DslValue'
import type { Arguments } from './Arguments'
import type { Environment } from './Environment'

export interface BuiltinFunction {
  name: string
  isDynamic?: boolean
  call: (args: Arguments, env: Environment) => DslValue | Promise<DslValue>
}

const functions = new Map<string, BuiltinFunction>()
let initialized = false

function ensureInitialized(): void {
  if (initialized) return
  initialized = true

  // Lazy imports to avoid circular dependency issues
  const { getDateFunctions } = require('../builtins/DateFunctions')
  const { getCharacterConstants } = require('../builtins/CharacterConstants')
  const { getArithmeticFunctions } = require('../builtins/ArithmeticFunctions')
  const { getComparisonFunctions } = require('../builtins/ComparisonFunctions')
  const { getPatternFunctions } = require('../builtins/PatternFunctions')
  const { getNoteFunctions } = require('../builtins/NoteFunctions')
  const { getListFunctions } = require('../builtins/ListFunctions')
  const { getSortConstants } = require('../builtins/SortConstants')
  const { getActionFunctions } = require('../builtins/ActionFunctions')

  const allFunctions: BuiltinFunction[] = [
    ...getDateFunctions(),
    ...getCharacterConstants(),
    ...getArithmeticFunctions(),
    ...getComparisonFunctions(),
    ...getPatternFunctions(),
    ...getNoteFunctions(),
    ...getListFunctions(),
    ...getSortConstants(),
    ...getActionFunctions(),
  ]

  for (const fn of allFunctions) {
    functions.set(fn.name, fn)
  }
}

export const BuiltinRegistry = {
  get(name: string): BuiltinFunction | null {
    ensureInitialized()
    return functions.get(name) ?? null
  },

  has(name: string): boolean {
    ensureInitialized()
    return functions.has(name)
  },

  isDynamic(name: string): boolean {
    ensureInitialized()
    return functions.get(name)?.isDynamic ?? false
  },

  allNames(): Set<string> {
    ensureInitialized()
    return new Set(functions.keys())
  },
}
